platform/pwl.h

```diff
@@ -18,9 +41,192 @@ typedef struct {
+    EGLContext context;
+    EGLConfig egl_config;
+} PwlEGLData;
+
+typedef struct {
+    struct xkb_context* context;
+    struct xkb_keymap* keymap;
+    struct xkb_state* state;
+
+    struct xkb_compose_table* compose_table;
+    struct xkb_compose_state* compose_state;
+
+    struct {
+        xkb_mod_index_t control;
+        xkb_mod_index_t alt;
+        xkb_mod_index_t shift;
+    } indexes;
+
+    struct {
+        xkb_mod_index_t control;
+        xkb_mod_index_t alt;
+        xkb_mod_index_t shift;
+    } modifier;
+
+    uint8_t modifiers;
+} PwlXKBData;
+
+#if HAVE_DEVICE_SCALING
+typedef struct output_metrics {
+  struct wl_output *output;
+  int32_t name;
+  int32_t scale;
+} output_metrics;
+#endif /* HAVE_DEVICE_SCALING */
+
+typedef struct {
+    struct wl_display *display;
+    struct wl_registry *registry;
+    struct wl_compositor *compositor;
+
+    struct xdg_wm_base *xdg_shell;
+    struct zwp_fullscreen_shell_v1 *fshell;
+    struct wl_shell *shell;
+
+    struct wl_seat *seat;
+
+#if HAVE_DEVICE_SCALING
+    struct output_metrics metrics[16];
+#endif /* HAVE_DEVICE_SCALING */
+
+    struct {
+        int32_t scale;
+    } current_output;
+
+    struct {
+        struct wl_pointer *obj;
+        int32_t x;
+        int32_t y;
+        uint32_t button;
+        uint32_t state;
+        struct wl_pointer_listener listener;
+    } pointer;
+
+    struct {
+        struct wl_keyboard *obj;
+
+        struct {
+            int32_t rate;
+            int32_t delay;
+        } repeat_info;
+
+        struct {
+            uint32_t key;
+            uint32_t time;
+            uint32_t state;
+            uint32_t event_source;
+        } repeat_data;
+
+        uint32_t serial;
+        struct wl_keyboard_listener listener;
+    } keyboard;
+
+    struct {
+        struct wl_touch *obj;
+        struct wl_touch_listener listener;
+    } touch;
+
+    GSource *event_src;
+
+    void (*resize_window)(void* data);
+    void (*handle_key_event)(void* data, uint32_t key, uint32_t state, uint32_t time);
+
+    struct wl_output_listener output_listener;
+} PwlData;
+
 PwlDisplay* pwl_display_connect (const char *name, GError**);
 void        pwl_display_destroy (PwlDisplay*);
 
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (PwlDisplay, pwl_display_destroy)
 
+#ifndef EGL_WL_create_wayland_buffer_from_image
+typedef struct wl_buffer * (EGLAPIENTRYP PFNEGLCREATEWAYLANDBUFFERFROMIMAGEWL) (EGLDisplay dpy, EGLImageKHR image);
+#endif
+
+struct pwl_event_source {
+    GSource source;
+    GPollFD pfd;
+    struct wl_display* display;
+};
+
+typedef struct {
+    struct wl_surface *wl_surface;
+    struct wl_surface_listener surface_listener;
+    struct wl_egl_window *egl_window;
+    EGLSurface egl_surface;
+
+    struct xdg_surface *xdg_surface;
+    struct xdg_toplevel *xdg_toplevel;
+    struct wl_shell_surface *shell_surface;
+
+    uint32_t width;
+    uint32_t height;
+
+    bool is_fullscreen;
+    bool is_maximized;
+} PwlWinData;
```

Comment:
This should be an opaque `PwlWindow` structure, and for each view we allocate one—instead of a global.

Reply:
The current implementation is reusing wayland surface. I suppose we could rework this staff restructuring the structures that you commented